node_exporter.go

```diff
@@ -188,10 +203,40 @@ func main() {
 			</html>`))
 	})
 
-	level.Info(logger).Log("msg", "Listening on", "address", *listenAddress)
-	server := &http.Server{Addr: *listenAddress}
-	if err := https.Listen(server, *configFile, logger); err != nil {
-		level.Error(logger).Log("err", err)
-		os.Exit(1)
+	var server *http.Server
+	if *socketPath == "" {
+		level.Info(logger).Log("msg", "Listening on", "address", *listenAddress)
+		server = &http.Server{Addr: *listenAddress}
+		go func() {
+			if err := https.Listen(server, *configFile, logger); err != nil {
+				level.Error(logger).Log("err", err)
+				os.Exit(1)
+			}
+		}()
+		<-done
+		level.Info(logger).Log("msg", "Connection closed", "address", *listenAddress)
+	} else {
+		level.Info(logger).Log("msg", "Listening unix socket on", "path", *socketPath)
+		server = &http.Server{}
+		os.Remove(*socketPath)
+		unixListener, err := net.Listen("unix", *socketPath)
+		if err != nil {
+			level.Error(logger).Log("err", err)
+			os.Exit(1)
+		}
+		if err := os.Chmod(*socketPath, os.FileMode(*socketPermissions)); err != nil {
+			level.Error(logger).Log("err", err)
+			os.Exit(1)
+		}
+		go func() {
+			if err := server.Serve(unixListener); err != nil {
+				level.Error(logger).Log("err", err)
+				os.Exit(1)
+			}
+		}()
+		<-done
+		level.Info(logger).Log("msg", "Connection closed", "path", *socketPath)
 	}
```

Comment:
Too much code duplication.

we can prepare human readable socket description separately, make logging lines the same and pull them out of "if" branches.

also, I would rename https.Listen to https.ConfigureServer and at the end, instead of doing ListenAndServeTLS at the end, just `return nil`. This way you can move go func() that runs the server also out of the "if".

In the end, I suggest something like:

	humanAddress := "address " + *listenAddress
	if *socketPath != "" {
		humanAddress = "path " + *socketPath
	}
	var server *http.Server
	var serve func() error
	level.Info(logger).Log("msg", "Listening on", humanAddress)
	server = &http.Server{}
	if *socketPath == "" {
		// tcpListener, err := ...
		if err := https.ConfigureServer(server, *configFile, logger); err != nil {
			level.Error(logger).Log("err", err)
			os.Exit(1)
		}
		serve = func() { return server.ServeTLS(tcpListener, "", "") }
	} else {
		os.Remove(*socketPath)
		unixListener, err := net.Listen("unix", *socketPath)
		if err != nil {
			level.Error(logger).Log("err", err)
			os.Exit(1)
		}
		if err := os.Chmod(*socketPath, os.FileMode(*socketPermissions)); err != nil {
			level.Error(logger).Log("err", err)
			os.Exit(1)
		}
		server = func() { return server.Serve(unixListener) }
	}
	go func() {
		if err := serve(); err != nil {
			level.Error(logger).Log("err", err)
			os.Exit(1)
		}
	}()
	<-done
	level.Info(logger).Log("msg", "Connection closed on", humanAddress)

Also, may be you should pull code that creates unix socket into separate getUnixListener() and you need to implement getTCPListener (there is a commented out line above)